File: app/src/main/java/com/github/gustavofernandes/radioplayer/MainActivity.java
package com.github.gustavofernandes.radioplayer;

import android.content.ComponentName;
import android.os.RemoteException;
import android.support.annotation.NonNull;
import android.support.v4.media.MediaBrowserCompat;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.widget.Button;
import android.widget.TextView;

import java.util.List;

public class MainActivity extends AppCompatActivity {

    private MediaBrowserCompat mMediaBrowser;

    private TextView mTitleTextView;
    private TextView mArtistTextView;
    private Button mPreviousButton;
    private Button mPlayPauseButton;
    private Button mNextButton;

    private boolean mIsPlaying;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mTitleTextView = findViewById(R.id.textview_title);
        mArtistTextView = findViewById(R.id.textview_artist);
        mPreviousButton = findViewById(R.id.button_previous);
        mPlayPauseButton = findViewById(R.id.button_playPause);
        mNextButton = findViewById(R.id.button_next);

        mMediaBrowser = new MediaBrowserCompat(
                this,
                new ComponentName(this, AudioService.class),
                new MediaBrowserConnectionCallback(),
                null);
    }

    @Override
    protected void onStart() {
        super.onStart();
        mMediaBrowser.connect();
    }

    private void buildTransportControls() {

        // TODO: make TransportControls function-wide?

        mPreviousButton.setOnClickListener(view -> MediaControllerCompat.getMediaController(MainActivity.this).getTransportControls().skipToPrevious());
        mNextButton.setOnClickListener(view -> MediaControllerCompat.getMediaController(MainActivity.this).getTransportControls().skipToNext());

        mPlayPauseButton.setOnClickListener(view -> {
            if (mIsPlaying) {
                MediaControllerCompat.getMediaController(MainActivity.this).getTransportControls().pause();
            } else {
                MediaControllerCompat.getMediaController(MainActivity.this).getTransportControls().play();
            }
        });

        updatePlayPauseView();
        updateTitleAndArtistViews(MediaControllerCompat.getMediaController(MainActivity.this).getMetadata());
    }

    private void updatePlayPauseView() {
        // TODO: account for states other than STATE_PLAYING or not
        mPlayPauseButton.setText(mIsPlaying ? "PAUSE" : "PLAY");
    }

    private void updateTitleAndArtistViews(MediaMetadataCompat metadata) {
        if (metadata != null) {
            mTitleTextView.setText(metadata.getString(MediaMetadataCompat.METADATA_KEY_TITLE));
            mArtistTextView.setText(metadata.getString(MediaMetadataCompat.METADATA_KEY_ARTIST));
        }
    }

    private class MediaBrowserConnectionCallback extends MediaBrowserCompat.ConnectionCallback {

        @Override
        public void onConnected() {
            try {

                MediaControllerCompat mediaController = new MediaControllerCompat(
                        MainActivity.this,
                        mMediaBrowser.getSessionToken());

                mediaController.registerCallback(new MediaControllerCallback());

                MediaControllerCompat.setMediaController(MainActivity.this, mediaController);

                buildTransportControls();

                mMediaBrowser.subscribe(AudioService.ROOT, new MediaBrowserSubscriptionCallback());

            } catch (RemoteException e) {
                // TODO: handle exception
                e.printStackTrace();
            }
        }
    }

    private class MediaBrowserSubscriptionCallback extends MediaBrowserCompat.SubscriptionCallback {

        @Override
        public void onChildrenLoaded(@NonNull String parentId, @NonNull List<MediaBrowserCompat.MediaItem> children) {
            // TODO
        }
    }

    private class MediaControllerCallback extends MediaControllerCompat.Callback {

        @Override
        public void onMetadataChanged(MediaMetadataCompat metadata) {
            updateTitleAndArtistViews(metadata);
        }

        @Override
        public void onPlaybackStateChanged(PlaybackStateCompat state) {
            mIsPlaying = state != null && state.getState() == PlaybackStateCompat.STATE_PLAYING;
            updatePlayPauseView();
        }
    }
}

File: app/src/main/java/com/github/gustavofernandes/radioplayer/MainActivity.kt
package com.github.gustavofernandes.radioplayer

import android.Manifest
import android.content.ComponentName
import android.content.pm.PackageManager
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v4.media.MediaBrowserCompat
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.support.v7.app.AppCompatActivity
import android.util.Log
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    private val TAG = MainActivity::class.java.simpleName

    private lateinit var mediaBrowser: MediaBrowserCompat

    private var isPlaying: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        Log.d(TAG, "onCreate")

        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        mediaBrowser = MediaBrowserCompat(
                this,
                ComponentName(this, AudioService::class.java),
                MediaBrowserConnectionCallback(),
                null)

        if(ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE), 0)
        }
    }

    override fun onStart() {
        Log.d(TAG, "onStart")

        super.onStart()

        Log.d(TAG, "Media browser connecting...")
        mediaBrowser.connect()
    }

    private fun buildTransportControls() {
        val mediaController = MediaControllerCompat.getMediaController(this@MainActivity)
        val transportControls = mediaController.transportControls

        button_previous.setOnClickListener { transportControls.skipToPrevious() }
        button_next.setOnClickListener { transportControls.skipToNext() }
        button_playPause.setOnClickListener {
            if (isPlaying) {
                Log.d(TAG, "Pause button clicked")
                transportControls.pause()
            } else {
                Log.d(TAG, "Play button clicked")
                transportControls.play()
            }
        }

        updatePlayPauseView()
        updateTitleAndArtistViews(mediaController.metadata)
    }

    private fun updatePlayPauseView() {
        // TODO: account for states other than STATE_PLAYING or not"
        button_playPause.text = if (isPlaying) "PAUSE" else "PLAY"
    }

    private fun updateTitleAndArtistViews(metadata: MediaMetadataCompat?) {
        if (metadata != null) {
            textview_title.text = metadata.getString(MediaMetadataCompat.METADATA_KEY_TITLE)
            textview_artist.text = metadata.getText(MediaMetadataCompat.METADATA_KEY_ARTIST)
        }
    }

    private fun loadStations() {
        Log.d(TAG, "Loading stations...")

        mediaBrowser.subscribe(AudioService.STATIONS, object : MediaBrowserCompat.SubscriptionCallback() {
            override fun onChildrenLoaded(parentId: String, children: MutableList<MediaBrowserCompat.MediaItem>) {
                Log.d(TAG, "Stations loaded: $children")

            }
        })
    }

    private inner class MediaBrowserConnectionCallback : MediaBrowserCompat.ConnectionCallback() {

        override fun onConnected() {
            Log.d(TAG, "Media browser connected")

            val mediaController = MediaControllerCompat(
                    this@MainActivity,
                    mediaBrowser.sessionToken)

            mediaController.registerCallback(MediaControllerCallback())

            MediaControllerCompat.setMediaController(this@MainActivity, mediaController)

            buildTransportControls()
            loadStations()
        }

        override fun onConnectionSuspended() {
            // TODO
        }

        override fun onConnectionFailed() {
            // TODO
        }
    }

    private inner class MediaBrowserSubscriptionCallback : MediaBrowserCompat.SubscriptionCallback() {

        override fun onChildrenLoaded(parentId: String, children: MutableList<MediaBrowserCompat.MediaItem>) {
            Log.d("MediaBrowserSubCallback", "Media browser onChildrenLoaded. parentId:$parentId, children:$children")
            // TODO
        }
    }

    private inner class MediaControllerCallback : MediaControllerCompat.Callback() {

        override fun onMetadataChanged(metadata: MediaMetadataCompat?) {
            updateTitleAndArtistViews(metadata)
        }

        override fun onPlaybackStateChanged(state: PlaybackStateCompat?) {
            isPlaying = state != null && state.state == PlaybackStateCompat.STATE_PLAYING
            updatePlayPauseView()
        }
    }
}
